src/webserver/DOSGuard.h

```diff
@@ -9,11 +9,48 @@
 class DOSGuard
 {
     boost::asio::io_context& ctx_;
-    std::mutex mtx_;  // protects ipFetchCount_
+    std::recursive_mutex
+        mtx_;  // protects ipFetchCount_, ipRequestCount_
+               // recursive mutex allows reentrant mutex acquisitions
+               // this makes isOk(ip) and checkout(ip) atomic
     std::unordered_map<std::string, std::uint32_t> ipFetchCount_;
+    std::unordered_map<std::string, std::uint32_t> ipRequestCount_;
     std::unordered_set<std::string> const whitelist_;
     std::uint32_t const maxFetches_;
     std::uint32_t const sweepInterval_;
+    std::uint32_t const maxConcurrentRequests_;
+
+    struct RPCScope
+    {
+        std::uint32_t& count_;
+        std::recursive_mutex& mtx_;
```

Comment:
Why does this need to be recursive?

Reply:
An RPCScope is constructed with a reference to the DOSGuard mutex. The choice of recursive mutex allows for reentrant mutex acquisitions, which allow sequences of function calls to be atomic. For example, isOk calls both getFetches and getRequests. These methods both acquire the mutex such to avoid data races. The problem is that if getFetches is called and then getRequests after, a different thread could perform some work in between the calls. The recursive mutex allows us to use a single lock instead of an ordered locking scheme.

Reply:
I'd rather avoid `recursive_mutex` as I feel it makes the concurrency pattern harder to reason about. Are `getFetches()` and `getRequests()` called anywhere besides `isOk()`?

Reply:
I agree with you here. I do not think `getFetches()` and `getRequests()` are called outside of `isOk()`. Currently, I would like to use a `shared_mutex` with `shared_lock` acquisitions in all of these methods. This would make `isOk()` atomic.